### PR TITLE
ornament (v5)

### PR DESCRIPTION
**Pull request by rvilarl**
Last class conversion to v5 before clean up.

### CODE REVIEW

**Comment by rvilarl** [comment]
Visual differences

current (BUGS: some ornaments are not below and some are not delayed. This is not a surprise there is no code to do that but I still did not find the right place to do it)
Ornament Jazz_Ornaments Bravura svg_current
reference
Ornament Jazz_Ornaments Bravura svg_reference
current
Ornament Ornaments Bravura svg_current
reference
Ornament Ornaments Bravura svg_reference
current
Ornament Ornaments___Delayed_turns Bravura svg_current
reference
Ornament Ornaments___Delayed_turns Bravura svg_reference
current
Ornament Ornaments_Vertically_Shifted Bravura svg_current
reference
Ornament Ornaments_Vertically_Shifted Bravura svg_reference
current
Ornament Stacked Bravura svg_current
reference
Ornament Stacked Bravura svg_reference
current
Ornament With_Upper_Lower_Accidentals Bravura svg_current
reference
Ornament With_Upper_Lower_Accidentals Bravura svg_reference

**Comment by ronyeh** [comment]
How was it done before, through xOffset and yOffset?

**Comment by rvilarl** [comment]
> How was it done before, through xOffset and yOffset?

I have reworked it to use the articulation approach, there is now only a small overlap between the delayed ornament and the accidental. I think that it is acceptable.

Ornament Jazz_Ornaments Bravura svg_current

**Comment by rvilarl** [comment]
@ronyeh ready to review! After this one there will be a lot of cleaning.

File: src/articulation.ts [hunk 1]
```diff
@@ -65,7 +65,7 @@ function snapLineToStaff(canSitBetweenLines: boolean, line: number, position: nu
 // Helper function for checking if a Note object is either a StaveNote or a GraceNote.
 const isStaveOrGraceNote = (note: Note) => isStaveNote(note) || isGraceNote(note);
 
-function getTopY(note: Note, textLine: number): number {
+export function getTopY(note: Note, textLine: number): number {
```
Comment (ronyeh) [inline src/articulation.ts:68]: OK, if these three exported functions are useful for both articulation and ornaments, maybe there should be a separate utility class to store them? We can figure this out in a separate PR.

File: src/fonts/common_metrics.ts [hunk 2]
```diff
@@ -1,75 +1,6 @@
...
-      reportedWidth: 31,
-    },
-    brassSmear: {
-      xOffset: 10,
-      yOffset: 0,
-      stemUpYOffset: 8,
-      reportedWidth: 5,
-    },
-  },
-
```
Comment (ronyeh) [inline src/fonts/common_metrics.ts:72]: I assume these were all hand-tuned at some point. If we need to improve the visual layout of the ornaments, we should figure out how to possibly reincorporate this visual hand tuning.

File: src/ornament.ts [hunk 3]
```diff
@@ -2,29 +2,22 @@
...
   if (Ornament.DEBUG) log('Vex.Flow.Ornament', args);
 }
 
-export interface OrnamentMetrics {
-  xOffset: number;
-  yOffset: number;
-  stemUpYOffset: number;
-  reportedWidth: number;
-}
-
```
Comment (ronyeh) [inline src/ornament.ts:27]: I like the way we have gone, with just using SMuFL glyphs directly, but perhaps there is some value to being able to hand tune the position of glyphs. Maybe later on we'll need to reintroduce little customizations like the xOffset / yOffset.

File: src/ornament.ts [hunk 4]
```diff
@@ -177,54 +142,50 @@ export class Ornament extends Modifier {
...
-    // new ornaments have their origin at the origin, and have more specific
-    // metrics.  Legacy ornaments do some
-    // x scaling, and have hard-coded metrics
-    const metrics = this.getMetrics();
-
     // some jazz ornaments are above or below depending on stem direction.
     this.adjustForStemDirection = false;
 
-    // some jazz ornaments like falls are supposed to overlap with future bars
-    // and so we report a different width than they actually take up.
```
Comment (ronyeh) [inline src/ornament.ts:200]: Interesting comment that is now deleted. :-) I am not an expert with jazz ornaments, so we'll have to get these reviewed by experts later on. How will we handle this overlap?

File: src/tables.ts [hunk 5]
```diff
@@ -427,33 +427,31 @@ const articulations: Record<string, ArticulationStruct> = {
...
-  fallLong: { code: 'brassFallRoughMedium' },
-  bend: { code: 'brassBend' },
-  plungerClosed: { code: 'brassMuteClosed' },
-  plungerOpen: { code: 'brassMuteOpen' },
-  flip: { code: 'brassFlip' },
-  jazzTurn: { code: 'brassJazzTurn' },
-  smear: { code: 'brassSmear' },
+const ornaments: Record<string, string> = {
+  mordent: '\ue56c' /*ornamentShortTrill*/,
+  mordentInverted: '\ue56d' /*ornamentMordent*/,
```
Comment (ronyeh) [inline src/tables.ts:432]: I thought Aaron had raised concerns about removing support for mordent_inverted and turn_inverted? If this breaks his app again we might have to reintroduce. :-)

File: src/tables.ts [hunk 6]
```diff
@@ -427,33 +427,31 @@ const articulations: Record<string, ArticulationStruct> = {
...
+  doit: '\ue5d5' /*brassDoitMedium*/,
+  fall: '\ue5d7' /*brassFallLipShort*/,
+  doitLong: '\ue5d2' /*brassLiftMedium*/,
+  fallLong: '\ue5de' /*brassFallRoughMedium*/,
+  bend: '\ue5e3' /*brassBend*/,
+  plungerClosed: '\ue5e5' /*brassMuteClosed*/,
+  plungerOpen: '\ue5e7' /*brassMuteOpen*/,
+  flip: '\ue5e1' /*brassFlip*/,
+  jazzTurn: '\ue5e4' /*brassJazzTurn*/,
+  smear: '\ue5e2' /*brassSmear*/,
```
Comment (ronyeh) [inline src/tables.ts:454]: It's interesting that some of these are camelcase, like turnInverted, and others are just all lowercase, like upmordent and downmordent....

**Review by ronyeh — COMMENTED** [review]
OK, I'll merge in the interest of moving forward. But I think we can revisit this class (with help from others) to improve the approach.

**Comment by rvilarl** [comment]
Thanks Ron! We will have to improve a lot of things, I am sure!

**Comment by codesenyou** [comment]
https://private-user-images.githubusercontent.com/22865285/266515291-2718e2d6-b923-4bcb-8e42-80b2b87326a2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc2NTM3NTQsIm5iZiI6MTc0NzY1MzQ1NCwicGF0aCI6Ii8yMjg2NTI4NS8yNjY1MTUyOTEtMjcxOGUyZDYtYjkyMy00YmNiLThlNDItODBiMmI4NzMyNmEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTE5VDExMTczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQyZTAzMTZhYWFkODlhZjAwYmE2M2YxN2M0M2RhMzJlZjFlNWI5NjUzMzZlY2U3NjRlMTIwZTA3Zjc2ZGMwNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IwD7qy8Vq_f9VCkAEY4lJLcJAss6V1kdJnjfvX-1V1E

How are you able to render the 6th, 8th and 10th element on this picture ?
I tried all the ornaments but I can only get those one

https://private-user-images.githubusercontent.com/22865285/266515293-916e37af-89ae-4da1-bec0-bb3dc9268c1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc2NTM3NTQsIm5iZiI6MTc0NzY1MzQ1NCwicGF0aCI6Ii8yMjg2NTI4NS8yNjY1MTUyOTMtOTE2ZTM3YWYtODlhZS00ZGExLWJlYzAtYmIzZGM5MjY4YzFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTE5VDExMTczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1NzQyMzNhNzk4NzNkNWRmOWI2N2U1OGEyYjc1MTMxNDA0MjE0ZDFhYjcxOTU0NGFjMDMwY2I4NTcxOWI3ODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KmAcR9bp1QblV7sA8Fs6YNNv4Ae_zeH7kdvMhKPc5N8

In order:
1) prallprall
2) prallprall (does inverted prallprall sign exist in vexflow ? It does on musescore)
3) downprall
4) upprall
5) upprall
6) downmordant
7) upmordant
8) pralldown
9) prallup
10) upprall
11) tr
12) mordent
13) mordent_inverted
14) turn
15) turn_inverted

This is what I am trying to render (capture from musescore)
image

**Pull request merged**